include/circt/Dialect/FIRRTL/FIRRTLStatements.td

```diff
@@ -228,6 +228,50 @@ def WhenOp : FIRRTLOp<"when", [SingleBlock, NoTerminator, NoRegionArguments,
   }];
 }
 
+def MatchOp : FIRRTLOp<"match", [SingleBlock, NoTerminator,
+    RecursiveMemoryEffects, RecursivelySpeculatable]> {
+  let summary = "Match Statement";
+  let description = [{
+    The "firrtl.match" operation represents a pattern matching statement on a
+    enumeration. This operation does not return a value and cannot be used as an
+    expression. Last connect semantics work similarly to a when statement.
+
+    Example:
+    ```mlir
+      firrtl.match %in : !firrtl.enum<Some: uint<1>, None: uint<0>> {
+        case Some(%arg0) {
```

Comment:
The `(` and `)` are not technically needed here, but I added them to match the FIRRTL syntax.

I might be able to get rid of the `case` keyword here but it made parsing easier.